File: app/src/main/kotlin/com/example/cranium/kernel/ReplayGuard.kt
package com.example.cranium.kernel

class InMemoryReplayGuard {
    data class Entry(
        val requestId: String,
        val idempotencyKey: String,
        val canonicalRequestHashHex: String,
        val transitionId: String,
        val timestamp: Long
    )

    private val entriesByRequestId = mutableMapOf<String, Entry>()
    private val entriesByIdempotencyKey = mutableMapOf<String, Entry>()
    private val transitionsById = mutableMapOf<String, AuthorityTransition>()

    fun inspect(
        requestId: String,
        idempotencyKey: String,
        canonicalRequestHash: RequestHash
    ): ReplayStatus {
        val byId = entriesByRequestId[requestId]
        val byKey = entriesByIdempotencyKey[idempotencyKey]

        if (byId == null && byKey == null) {
            return ReplayStatus.New
        }

        if (byId != null && byId.idempotencyKey == idempotencyKey && byId.canonicalRequestHashHex == canonicalRequestHash.hexDigest) {
            return ReplayStatus.Existing(byId.transitionId, byId.timestamp)
        }

        if (byKey != null && byKey.requestId == requestId && byKey.canonicalRequestHashHex == canonicalRequestHash.hexDigest) {
            return ReplayStatus.Existing(byKey.transitionId, byKey.timestamp)
        }

        val conflictReason = if (byId != null) {
            "RequestId '$requestId' previously used with hash '${byId.canonicalRequestHashHex}' (attempted '${canonicalRequestHash.hexDigest}')"
        } else {
            "IdempotencyKey '$idempotencyKey' previously used with hash '${byKey?.canonicalRequestHashHex}' (attempted '${canonicalRequestHash.hexDigest}')"
        }

        return ReplayStatus.ConflictingReuse(
            reason = conflictReason,
            priorHash = byId?.canonicalRequestHashHex ?: byKey?.canonicalRequestHashHex ?: "",
            attemptedHash = canonicalRequestHash.hexDigest
        )
    }

    fun record(
        requestId: String,
        idempotencyKey: String,
        canonicalRequestHash: RequestHash,
        transition: AuthorityTransition
    ) {
        val entry = Entry(
            requestId = requestId,
            idempotencyKey = idempotencyKey,
            canonicalRequestHashHex = canonicalRequestHash.hexDigest,
            transitionId = transition.id,
            timestamp = System.currentTimeMillis()
        )
        entriesByRequestId[requestId] = entry
        entriesByIdempotencyKey[idempotencyKey] = entry
        transitionsById[transition.id] = transition
    }

    fun loadCommittedTransition(transitionId: String): AuthorityTransition? = transitionsById[transitionId]

    fun getAllEntries(): List<Entry> = entriesByRequestId.values.toList()

    fun clear() {
        entriesByRequestId.clear()
        entriesByIdempotencyKey.clear()
        transitionsById.clear()
    }
}

File: src/kernel/replayGuard.ts
import { ReplayStatus } from './types';

export interface ReplayRecord {
  idempotencyKey: string;
  canonicalRequestHashHex: string;
  transitionId: string;
  timestamp: number;
}

export class InMemoryReplayGuard {
  private records = new Map<string, ReplayRecord>();

  check(idempotencyKey: string, canonicalHashHex: string): ReplayStatus {
    const existing = this.records.get(idempotencyKey);
    if (!existing) {
      return { type: 'New' };
    }

    if (existing.canonicalRequestHashHex === canonicalHashHex) {
      return {
        type: 'Existing',
        transitionId: existing.transitionId,
        cachedTimestamp: existing.timestamp,
      };
    }

    return {
      type: 'ConflictingReuse',
      reason: `Idempotency key '${idempotencyKey}' was previously bound to hash ${existing.canonicalRequestHashHex.slice(0, 16)}..., cannot reuse with different hash ${canonicalHashHex.slice(0, 16)}...`,
      priorHash: existing.canonicalRequestHashHex,
      attemptedHash: canonicalHashHex,
    };
  }

  record(record: ReplayRecord): void {
    this.records.set(record.idempotencyKey, record);
  }

  getAllEntries(): ReplayRecord[] {
    return Array.from(this.records.values());
  }

  clear(): void {
    this.records.clear();
  }
}
